yente/data/delta.py

```diff
@@ -0,0 +1,108 @@
+from typing import Optional, TypedDict, Dict, List, Any
+from typing import AsyncGenerator, Tuple
+
+from yente.logs import get_logger
+from yente.data.dataset import Dataset
+from yente.data.loader import load_json_url, load_json_lines
+
+log = get_logger(__name__)
+
+
+class DeltaIndex(TypedDict):
+    versions: Dict[str, str]
+
+
+class EntityOp(TypedDict):
+    op: str
+    entity: Dict[str, Any]
+
+
+class DatasetLoader(object):
+    """A helper object for emitting entity operations to transition from one
+    loaded dataset version to the next."""
+
+    def __init__(
+        self, dataset: Dataset, base_version: Optional[str], force_full: bool = False
+    ) -> None:
+        self.dataset = dataset
+        self.target_version = dataset.version or "static"
+        self.base_version = base_version
+        self.force_full = force_full
+        self.delta_urls: Optional[List[Tuple[str, str]]] = None
+
+    @classmethod
+    async def build(
+        cls, dataset: Dataset, base_version: Optional[str], force_full: bool = False
+    ) -> "DatasetLoader":
+        """Fetch the index of delta files and decide an index building strategy."""
+        loader = DatasetLoader(dataset, base_version, force_full=force_full)
+        if 
+        if dataset.delta_url is None:
+            log.debug("No delta updates available for: %r" % dataset.name)
+            return loader
+        if loader.base_version is None or loader.target_version <= loader.base_version:
+            return loader
+
+        index: DeltaIndex = await load_json_url(dataset.delta_url)
+        versions = index.get("versions", {})
+        sorted_versions = sorted(versions.keys())
+        if len(sorted_versions) == 0:
+            return loader
+        if loader.base_version not in sorted_versions:
+            log.warning(
+                "Loaded version of dataset is older than delta window",
+                dataset=dataset.name,
+                delta_url=dataset.delta_url,
+                base_version=loader.base_version,
+                target_version=loader.target_version,
+                delta_versions=sorted_versions,
+            )
+            return loader
+
+        loader.delta_urls = []
+        for version in sorted_versions:
+            if version <= loader.base_version or version > loader.target_version:
+                continue
+            loader.delta_urls.append((version, versions[version]))
+
+        # TODO: is this smart? this avoids running clones when there is not change:
+        loader.target_version = max(sorted_versions)
+        return loader
+
+    @property
+    def is_incremental(self) -> bool:
+        """Check if there is sequence of delta entity patches that can be loaded."""
+        if self.force_full:
+            return False
+        return self.delta_urls is not None and len(self.delta_urls) > 0
+
+    def check(self) -> bool:
```

Comment:
Rename to `needs_update`? `check` makes it a bit hard to parse from the callsite.

Reply:
Very very true!